server/coordinator/procedure/manager_impl.go

```diff
@@ -123,34 +121,44 @@ func (m *ManagerImpl) retryAll(ctx context.Context) error {
 	return nil
 }
 
-func (m *ManagerImpl) startProcedureWorker(ctx context.Context, procedures <-chan Procedure) {
-	for procedure := range procedures {
-		err := procedure.Start(ctx)
+func (m *ManagerImpl) startProcedurePromote(ctx context.Context) {
+	for {
+		err, newProcedures := m.promoteProcedure(ctx)
 		if err != nil {
-			log.Error("procedure start failed", zap.Error(err))
+			log.Error("promote procedure failed", zap.Error(err))
+			continue
+		}
+		if len(newProcedures) > 0 {
```

Comment:
Is this check necessary?

Reply:
The check here is necessary because promote may return an empty array when the queue is empty.